src/device_memory_tracker.cc

```diff
@@ -0,0 +1,288 @@
+       ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) \
+       : (buffer))
+
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
+}
+
+}  // namespace
+
+DeviceMemoryTracker::DeviceMemoryTracker()
+{
+  cudaError_t cuerr = cudaGetDeviceCount(&device_cnt_);
+  if ((cuerr == cudaErrorNoDevice) || (cuerr == cudaErrorInsufficientDriver)) {
+    device_cnt_ = 0;
+  } else if (cuerr != cudaSuccess) {
+    throw std::runtime_error("Unexpected failure on getting CUDA device count.");
+  }
+
+  // Use 'cuptiSubscribe' to check if the cupti has been initialized
+  // elsewhere. Due to cupti limitation, there can only be one cupti client
+  // within the process, so in the case of per-backend memory tracking, we
+  // have to make the assumption that the other cupti client is using the same
+  // memory tracker implementation so that the backend may use the cupti
+  // configuration that is external to the backend without issue.
+  auto cupti_res = cuptiSubscribe(&subscriber_, nullptr, nullptr);
+  switch (cupti_res)
+  {
+  case CUPTI_SUCCESS : {
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
+    break;
+  }
+  default:
+    // other error, should propagate and disable memory tracking for the backend
+    throw std::runtime_error("Unexpected failure on configuring CUPTI.");
+  }
+}
+
+int
+DeviceMemoryTracker::CudaDeviceCount()
+{
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  return 0;
```

Comment:
can we add `throw std::runtime_error("DeviceMemoryTracker::Init() must be called before using any DeviceMeoryTracker features.");`

Reply:
No, this is not throw intentionally, `MemoryUsage` will access this function and we will have to do a lot more handling if the backend disable the tracking.

Reply:
but the ordering is important - i.e. if you have a MemoryUsage object that gets created before Init() - it will have the wrong device count - so need someway to catch that -

Reply:
That only happen if you intentionally do so. Again, Init() is called on BackendInitiaze() which is the first place that the backend code will be executed. MemoryUsage is tracking usage of the model / instance which happens much later. I don't think it is common user error to initialize MemoryUsage object on BackendInitiaze, or actually not calling Init() as the first action when using the utility. I can see the user may forget to call `Init()` entirely, then the exception will be thrown anyway when they are calling the track function.

Reply:
what happens if we have the throw? Trying to understand - this is so we can use memory usage without calling init()? (so calls can be left in?)

Reply:
Actually I think we can throw, I have changed the implementation locally to handle disabled case without too much conditionals and in such case the memory usage object won't be initialized.